Implement Rust native article view as fallback on failed `article_parse_command`

0x6b opened the pull request:
First of all, thank you for bringing this great TUI for us!

### Summary

Implement Rust native article view as fallback on failed `article_parse_command`.

### Context

Current implementation will display a link to the [`config.md`](https://github.com/aome510/hackernews-TUI/blob/main/docs/config.md#article-parse-command) file if the `article_parse_command` fails. Almost all of new users will see this error. Instead, we can use the `readable-readability` to provide similar capability as a fallback method. This might be more user friendly and nice default behavior. 

### Dependency Added

- [`readable-readability`](https://crates.io/crates/readable-readability)

### Screenshots

https://news.ycombinator.com/item?id=37214693

|`article_md`|New implementation|
|-|-|
|![article_md 1](https://github.com/aome510/hackernews-TUI/assets/679719/a85e6f5b-d003-44c4-9270-72555535631d)|![integrated 1](https://github.com/aome510/hackernews-TUI/assets/679719/5edcc5d8-d339-431b-a331-90d51833df46)|

Please note that it's not perfect for some pages as below, but I think it'd be acceptable.

https://news.ycombinator.com/item?id=25607320

|`article_md`|New implementation|
|-|-|
|![aritcle_md 2](https://github.com/aome510/hackernews-TUI/assets/679719/87702be2-aa88-4169-8199-f83e677c761a)|![integrated 2](https://github.com/aome510/hackernews-TUI/assets/679719/1302615c-058b-4277-a97a-4d06ff439550)|

### Final Words

Quick and not-so-scientific benchmarking shows that the native implementation is faster than the `article_md`, separate Node.js CLI, on my setup. This is just a proposal. I'm not sure if it's a good idea to add a new dependency to the project. Feel free to close this PR if you don't like it. If you like it, I'd add commits to update relevant documentation. Thank you!

aome510 commented on hackernews_tui/src/client/mod.rs:
```diff
@@ -376,27 +376,65 @@ impl HNClient {
+                    .map_err(|why| anyhow::anyhow!("failed to turn the response into string: {why}"))?;
+                let (nodes, metadata) = readable_readability::Readability::new()
+                    .base_url(url::Url::parse(url)
+                        .map_err(|why| anyhow::anyhow!("failed to parse url: {why}"))?)
```
can use `anyhow::Context`, e.g `.context("failed to parse url")` instead.

0x6b replied:
Addressed at [1d03f33](https://github.com/aome510/hackernews-TUI/pull/94/commits/1d03f33a403a36ded9f07de77d7e52186148f376).

aome510 commented on hackernews_tui/src/client/mod.rs:
```diff
@@ -376,27 +376,65 @@ impl HNClient {
-            Err(anyhow::anyhow!(stderr))
+            Err(_) => {
+                // fallback to the `readable-readability` crate if the command fails
+                let html = ureq::AgentBuilder::new()
+                    .timeout(std::time::Duration::from_secs(config::get_config().client_timeout))
+                    .build()
+                    .get(url)
```
This is a function inside `HNClient` which has a ureq client already. Can you update the function signature to take `&self` and use `self.client` instead?

0x6b replied:
Addressed at [28d89b5](https://github.com/aome510/hackernews-TUI/pull/94/commits/28d89b5a9f9394a8a5d0721f401363b04e8d6bfc)

aome510 commented on hackernews_tui/src/client/mod.rs:
```diff
@@ -376,27 +376,65 @@ impl HNClient {
+                    .build()
+                    .get(url)
+                    .call()
+                    .map_err(|why| anyhow::anyhow!("failed to get url: {why}"))?
+                    .into_string()
+                    .map_err(|why| anyhow::anyhow!("failed to turn the response into string: {why}"))?;
```
can use `anyhow::Context` instead

0x6b replied:
Addressed at [1d03f33](https://github.com/aome510/hackernews-TUI/pull/94/commits/1d03f33a403a36ded9f07de77d7e52186148f376).

aome510 commented on hackernews_tui/src/client/mod.rs:
```diff
@@ -376,27 +376,65 @@ impl HNClient {
+
+                let mut text = vec![];
+                nodes.serialize(&mut text)
+                    .map_err(|why| anyhow::anyhow!("failed to serialize nodes: {why}"))?;
```
can use `anyhow::Context` instead

0x6b replied:
Addressed at [1d03f33](https://github.com/aome510/hackernews-TUI/pull/94/commits/1d03f33a403a36ded9f07de77d7e52186148f376).

aome510 commented on hackernews_tui/src/client/mod.rs:
```diff
@@ -376,27 +376,65 @@ impl HNClient {
+                    .or(metadata.article_title)
+                    .unwrap_or("(no title)".to_string());
+                let content = std::str::from_utf8(&text)
+                    .map_err(|why| anyhow::anyhow!("failed to turn the text into string: {why}"))?
```
can use `anyhow::Context` instead

0x6b replied:
Addressed at [1d03f33](https://github.com/aome510/hackernews-TUI/pull/94/commits/1d03f33a403a36ded9f07de77d7e52186148f376).

aome510 reviewed (changes requested):
> Quick and not-so-scientific benchmarking shows that the native implementation is faster than the article_md, separate Node.js CLI, on my setup. This is just a proposal. I'm not sure if it's a good idea to add a new dependency to the project. Feel free to close this PR if you don't like it. If you like it, I'd add commits to update relevant documentation. Thank you!

Looking good. Thank you for this PR. Can you update the documentation in https://github.com/aome510/hackernews-TUI/blob/main/docs/config.md#article-parse-command and also the error message in https://github.com/aome510/hackernews-TUI/blob/bfc9e7f01bfb9a268d9826c88e764c183fa5c699/hackernews_tui/src/view/async_view.rs#L55-L61?

0x6b commented:
@aome510 Thank you for the detailed review. I have addressed your comments. I think some updated wordings should be corrected per your liking, please treat as a start point. Welcome any further feedback. Thanks!

0x6b commented:
@aome510 Thanks! I should have applied `cargo fmt` before pushing commits, of course. Please review my recent commits and let me know if you need anything else.

aome510 commented:
Great! Thanks @0x6b
